docs/rfc/0004_component_groups_scale_view.md

```diff
@@ -0,0 +1,300 @@
+
+The [scale subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource)
+is
+a subresource like the status but exposes three fixed fields:
+
+- `.spec.replicas` which refers to the desired replicas
+- `.status.replicas` which refers to the current replicas
+- `.status.selector` which refers to the Pod selector (but in a serialized manner)
+
+However, the current design of the `RisingWave` resource can not provide such a subresource because it is consisted of
+four independent components, and each of the component can have one or more groups. Therefore, the `kubectl scale` can
+not work directly on the `RisingWave` resource and so do other tools depend on that. And it is the main reason for
+proposing this new CR. I named it 'RisingWaveScaleView'.
+
+> What use cases does it support?
+
+With the `RisingWaveScaleView`, the users can set up a view (similar to the view in databases) of some groups
+of one RisingWave component, in a `RisingWave` object, to expose the scale subresource. Then the magic happens and
+everything that depends on the scale subresource shall work.
+
+Take the auto-scaler as an example, we can set up the HPA or KEDA for frontend Pods with
+a `RisingWaveScaleView` that targets the frontend groups.
+
+> What is the expected outcome?
+
+1. A CR `RisingWaveScaleView` and its controller, with the following features:
+    - Users can target any of the four components in it, but aren't able to change it after.
+    - Users can target one or more valid groups in it, but aren't able to change it after.
+    - Users can set the priorities of groups for scaling out and in separately.
+    - It reflects all the fields required by the scale subresource described above and keeps them synced with the
+      RisingWave's.
+    - It locks the target RisingWave object on the specified groups and prevents the direct modifications (either update
+      or delete) on the related fields.
+        - It acquires the lock at the creation and removes it before deletion.
+
+2. An improvement on `RisingWave` resource and validating webhook to support lock of groups:
+    - Lock fields in the status, which is structured and easy to lookup.
+    - Enhanced validating webhook to reject any changes from besides the controller of `RisingWaveScaleView`.
+
+# **Explanation**
+
+> This is the technical portion of the RFC. Explain the design in sufficient detail.
+
+## `RisingWaveScaleView` CRD
+
+```yaml
+apiVersion: risingwave.risingwavelabs.com/v1alpha1
+kind: RisingWaveScaleView
+metadata:
+  name: hello-frontend-scale-view
+spec:
+  targetRef:
+    name: hello             # Name of the RisingWave object, in the same namespace.
+    component: frontend     # One of meta, frontend, compute, and compactor.
+  replicas: 10              # Target replicas. It will be synced from the RisingWave object when the controller grabs the lock.
+  # Strict verification mode on replicas:
```

Comment:
Looks like we should always enforce strict mode.

Reply:
I agree.